src/Features/Core/Portable/Diagnostics/Analyzers/QualifyMemberAccessDiagnosticAnalyzerBase.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.Diagnostics.QualifyMemberAccess
+{
+    internal abstract class QualifyMemberAccessDiagnosticAnalyzerBase<TLanguageKindEnum> : DiagnosticAnalyzer, IBuiltInAnalyzer where TLanguageKindEnum : struct
+    {
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(WorkspacesResources.MemberAccessShouldBeQualified), WorkspacesResources.ResourceManager, typeof(WorkspacesResources));
+
+        private static readonly LocalizableString s_localizableTitleQualifyMembers = new LocalizableResourceString(nameof(FeaturesResources.AddThisOrMeQualification), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+        private static readonly DiagnosticDescriptor s_descriptorQualifyMemberAccess = new DiagnosticDescriptor(IDEDiagnosticIds.AddThisOrMeDiagnosticId,
+                                                                    s_localizableTitleQualifyMembers,
+                                                                    s_localizableMessage,
+                                                                    DiagnosticCategory.Style,
+                                                                    DiagnosticSeverity.Hidden,
+                                                                    isEnabledByDefault: true,
+                                                                    customTags: DiagnosticCustomTags.Unnecessary);
+
+        private OptionSet _lazyDefaultOptionSet;
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+        {
+            get
+            {
+                return ImmutableArray.Create(s_descriptorQualifyMemberAccess);
+            }
+        }
+
+        protected abstract ImmutableArray<TLanguageKindEnum> GetSupportedSyntaxKinds();
+        protected abstract string GetLanguageName();
+        protected abstract bool IsCandidate(SyntaxNode node);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeNode, GetSupportedSyntaxKinds());
+        }
+
+        private void AnalyzeNode(SyntaxNodeAnalysisContext context)
+        {
+            if (!IsCandidate(context.Node))
```

Comment:
Actually, i think what i'm talking about already exists:

```c#
    /// <summary>
    /// Represents a reference to a C# this or VB Me parameter.
    /// </summary>
    public interface IInstanceReferenceExpression : IParameterReferenceExpression
    {
        /// <summary>
        /// Indicates whether the reference is explicit or implicit in source.
        /// </summary>
        bool IsExplicit { get; }
    }
```

So we already have an IOperation type that tells you you're referencing something on the instance.  *and* that node will also tell you if the 'this/me' is explicit or implicit.  This seems like precisely what we want to be using here.  

With this we should not need any language specific code for this feature.